petab_select/candidate_space.py

```diff
@@ -529,6 +602,628 @@ def iterate():
+
+        logging.info("JUMPING: ", new_init_model.parameters)
+        # breakpoint()
+
+        # if model not appropriate make it so by adding the first
+        # critical parameter from each critical parameter set
+        if not self.check_critical(new_init_model):
+            for critical_set in self.critical_parameter_sets:
+                new_init_model.parameters[critical_set[0]] = ESTIMATE
+
+        self.update_method(self.initial_method)
+
+        self.number_of_reattempts -= 1
+        self.jumped_to_most_distant = True
+
+        self.predecessor_model = new_init_model
+        self.best_model_of_current_run = new_init_model
+        if self.summary_tsv is not None:
+            with open(self.summary_tsv, 'a') as f:
+                writer = csv.writer(f, delimiter='\t')
+                writer.writerow(
+                    [
+                        "Jumped to the most distant model.",
+                        *([''] * 5),
+                    ]
+                )
+
+    # TODO Fix for non-famos model subspaces. FAMOS easy beacuse of only 0;ESTIMATE
+    def get_most_distant(self) -> Model:
+        """
+        Get most distant model to all the checked models. We take models from the
+        sorted list of best models (self.best_models) and construct complements of
+        these models. For all these complements we compute the distance in number of
+        different estimated parameters to all models from history. For each complement
+        we take the minimum of these distances as it's distance to history. Then we
+        choose the complement model with the maximal distance to history.
+        TODO:
+        Next we check if this model is contained in any subspace. If so we choose it.
+        If not we choose the model in a subspace that has least distance to this
+        complement model.
+        """
+
+        most_distance = 0
+        most_distant_indices = []
+
+        for model in self.best_models:
+            model_parameter_ids = list(model.petab_parameters)
+            model_parameters = model.get_parameter_values(
+                parameter_ids=model_parameter_ids
+            )
+
+            model_estimated_parameters = np.array(
+                [p == ESTIMATE for p in model_parameters]
+            ).astype(int)
+            complement_parameters = abs(model_estimated_parameters - 1)
+            # initialize the least distance to the maximal possible value of it
+            complement_least_distance = len(complement_parameters)
+            # get the complement least distance
+            for history_model_id in self.history:
+                history_model = self.history[history_model_id]
+                history_model_parameter_ids = list(
+                    history_model.petab_parameters
+                )
+                history_model_parameters = history_model.get_parameter_values(
+                    parameter_ids=history_model_parameter_ids
```

Comment:
Should be `model_parameter_ids` right? To ensure comparable vectors?

Reply:
I've replaced in the whole function all petab_parameter_ids with only one from self.best_models[0].petab_parameters